### PR TITLE
TabixReader : delete s.toCharArray()

### PR DESCRIPTION
**Pull request by lindenb**
### Description

the main motivation of this small PR was to remove the  line.toCharArray() in TabixReader https://github.com/samtools/htsjdk/blob/master/src/main/java/htsjdk/tribble/readers/TabixReader.java#L419

```
                    char[] str = s.toCharArray();
(...)
                    if (str.length == 0 || str[0] == mMeta) continue;
```

replaced with:

```
      if (s.isEmpty() || s.charAt(0) == mMeta) continue;
```

while I was here, I've added a few  `final`, fixed the name of some variables (e.g: 's' -> 'contig').

I've also added a variable `longestLineLength` that is used to allocate the capacity in the function:

```
    private static String readLine(final InputStream is,final int bufferCapacity) throws IOException {
```

### Checklist

- [x] Code compiles correctly
- [ ] New tests covering changes and new functionality
- [x] All tests passing
- [x] Extended the README / documentation, if necessary
- [ ] Is not backward compatible (breaks binary or source compatibility)

### CODE REVIEW

File: src/main/java/htsjdk/tribble/readers/TabixReader.java [hunk 1]
```diff
@@ -533,6 +554,8 @@ public static void main(String[] args) {
                     System.out.println(s);
             }
```
Comment (lbergelson) [inline src/main/java/htsjdk/tribble/readers/TabixReader.java]: This main method is bad and untested.  Instead of trying to fix it can you just delete it?

File: src/main/java/htsjdk/tribble/readers/TabixReader.java [hunk 2]
```diff
@@ -414,11 +435,11 @@ public String next() throws IOException {
                     ++i;
                 }
                 String s;
-                if ((s = readLine(mFp)) != null) {
+                if ((s = readLine(mFp,TabixReader.this.longestLineLength)) != null) {
+                    TabixReader.this.longestLineLength = Math.max(s.length(), TabixReader.this.longestLineLength);
                     TIntv intv;
-                    char[] str = s.toCharArray();
```
Comment (lbergelson) [inline src/main/java/htsjdk/tribble/readers/TabixReader.java]: this is a good change, that's a waste of allocation

File: src/main/java/htsjdk/tribble/readers/TabixReader.java [hunk 3]
```diff
@@ -56,6 +56,9 @@
     private int mBc;
     private int mEc;
     private int mMeta;
+    /** longest line seen so far, used to allocate the capacity of the lines */
+    private int longestLineLength = 100;
```
Comment (lbergelson) [inline src/main/java/htsjdk/tribble/readers/TabixReader.java]: Could you separate out this change to the longest line?  It seems orthogonal to the rest of the changes.  I see how it could potentially save us some allocations, but it adds complexity with somewhat questionable return.  If you want to change it to just set the default buffer size to something greater than 16, we could merge that as part of this PR.  Otherwise lets do that in a separate PR.

File: src/main/java/htsjdk/tribble/readers/TabixReader.java [hunk 4]
```diff
@@ -189,33 +192,47 @@ public static int readInt(final InputStream is) throws IOException {
...
     public static long readLong(final InputStream is) throws IOException {
-        byte[] buf = new byte[8];
+        final byte[] buf = new byte[8];
         is.read(buf);
         return ByteBuffer.wrap(buf).order(ByteOrder.LITTLE_ENDIAN).getLong();
     }
 
     public static String readLine(final InputStream is) throws IOException {
-        StringBuffer buf = new StringBuffer();
+        return readLine(is,100);
```
Comment (yfarjoun) [inline src/main/java/htsjdk/tribble/readers/TabixReader.java]: space after `,`

File: src/main/java/htsjdk/tribble/readers/TabixReader.java [hunk 5]
```diff
@@ -189,33 +192,47 @@ public static int readInt(final InputStream is) throws IOException {
...
+
+    /** 
+     * reads a line with a defined buffer-size 
+     * 
+     * @param is the input stream
+     * @param bufferCapacity the buffer size, must be greater than 0
+     * @return the line or null if there is no more input
+     * @throws IOException
+     */
+    private static String readLine(final InputStream is,final int bufferCapacity) throws IOException {
```
Comment (yfarjoun) [inline src/main/java/htsjdk/tribble/readers/TabixReader.java]: space after `, `

File: src/main/java/htsjdk/tribble/readers/TabixReader.java [hunk 6]
```diff
@@ -266,20 +283,24 @@ private void readIndex(SeekableStream fp) throws IOException {
...
+        final ISeekableStreamFactory ssf = SeekableStreamFactory.getInstance();
         readIndex(ssf.getBufferedStream(ssf.getStreamFor(mIndexPath, mIndexWrapper), 128000));
     }
 
     /**
      * Read one line from the data file.
      */
     public String readLine() throws IOException {
-        return readLine(mFp);
+        final String line = readLine(mFp,this.longestLineLength);
```
Comment (yfarjoun) [inline src/main/java/htsjdk/tribble/readers/TabixReader.java]: space and pull back the `}`

File: src/main/java/htsjdk/tribble/readers/TabixReader.java [hunk 7]
```diff
@@ -266,20 +283,24 @@ private void readIndex(SeekableStream fp) throws IOException {
...
+        if (line != null) {
+            this.longestLineLength = Math.max(this.longestLineLength, line.length());
+            }
+        return line;
     }
 
     /** return chromosome ID or -1 if it is unknown */
     public int chr2tid(final String chr) {
-        Integer tid=this.mChr2tid.get(chr);
+       final Integer tid=this.mChr2tid.get(chr);
```
Comment (yfarjoun) [inline src/main/java/htsjdk/tribble/readers/TabixReader.java]: while you are here, spaces around ` = `

File: src/main/java/htsjdk/tribble/readers/TabixReader.java [hunk 8]
```diff
@@ -414,11 +435,11 @@ public String next() throws IOException {
                     ++i;
                 }
                 String s;
-                if ((s = readLine(mFp)) != null) {
+                if ((s = readLine(mFp,TabixReader.this.longestLineLength)) != null) {
```
Comment (yfarjoun) [inline src/main/java/htsjdk/tribble/readers/TabixReader.java]: space

File: src/main/java/htsjdk/tribble/readers/TabixReader.java (hunk 6 above)
Comment (yfarjoun) [inline src/main/java/htsjdk/tribble/readers/TabixReader.java]: this function is untested.

File: src/main/java/htsjdk/tribble/readers/TabixReader.java (hunk 4 above)
Comment (yfarjoun) [reply src/main/java/htsjdk/tribble/readers/TabixReader.java]: also, this function is untested

**Comment by lindenb** [comment]
@lbergelson @yfarjoun back to you.

* I've removed the '`longest_line`'  but I've declared a ` static final int DEFAULT_BUFFER_SIZE = 1000` (to be moved to Defaults.java ) ?
* I've removed the main method
* I've added a test for TabixReader.readLine() but I'm not sure if it's what you wanted.

**Comment by lbergelson** [comment]
@lindenb Thanks you!
